storage/paths/db_index.go

```diff
@@ -0,0 +1,986 @@
+		for _, rft := range read.AllSet() {
+			if isLocked(lockMap[rft].writeTs) {
+				return false, xerrors.Errorf("cannot acquire read lock for sector %v filetype %d already locked for writing", sector, rft)
+			}
+		}
+
+		// Acquire write locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET write_ts = NOW(), write_lock_owner = $1
+							 	WHERE miner_id=$2
+							 		AND sector_num=$3
+							 		AND sector_filetype = ANY($4)`,
+			lockUuid.String(),
+			sector.Miner,
+			sector.Number,
+			write.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring write locks for sector %v fails with err: %v", sector, err)
+		}
+
+		// Acquire read locks
+		_, err = dbi.harmonyDB.Exec(ctx,
+			`UPDATE sectorlocation
+							 	SET read_ts = NOW(), read_refs = read_refs + 1
+							 	WHERE miner_id=$1
+							 		AND sector_num=$2
+							 		AND sector_filetype = ANY($3)`,
+			sector.Miner,
+			sector.Number,
+
+			read.AllSet())
+		if err != nil {
+			return false, xerrors.Errorf("acquiring read locks for sector %v fails with err: %v", sector, err)
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (dbi *DBIndex) unlock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType, lockUuid uuid.UUID) (bool, error) {
+	if read|write == 0 {
+		return false, nil
+	}
+
+	if read|write > (1<<storiface.FileTypes)-1 {
+		return false, xerrors.Errorf("unknown file types specified")
+	}
+
+	// Relinquish write locks
+	_, err := dbi.harmonyDB.Exec(ctx,
+		`UPDATE sectorlocation
+				SET write_ts = NULL, write_lock_owner = NULL
+				WHERE miner_id=$2
+					AND sector_num=$3
+					AND write_lock_owner=$4
+					AND sector_filetype = ANY($5)`,
+		lockUuid.String(),
+		sector.Miner,
+		sector.Number,
+		lockUuid.String(),
+		write.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing write locks for sector %v fails with err: %v", sector, err)
+	}
+
+	// Relinquish read locks
+	_, err = dbi.harmonyDB.Exec(ctx,
+		`UPDATE sectorlocation
+				SET read_refs = read_refs-1,
+					read_ts = CASE WHEN read_refs - 1 = 0 THEN NULL ELSE read_ts END
+				WHERE miner_id=$1
+					AND sector_num=$2
+					AND sector_filetype = ANY($3)`,
+		sector.Miner,
+		sector.Number,
+		read.AllSet())
+	if err != nil {
+		return false, xerrors.Errorf("relinquishing read locks for sector %v fails with err: %v", sector, err)
+	}
+
+	return true, nil
+
+}
+
+func (dbi *DBIndex) StorageLock(ctx context.Context, sector abi.SectorID, read storiface.SectorFileType, write storiface.SectorFileType) error {
+
+	retries := 5
+	waitTime := 1
+
+	// generate uuid for this lock owner
+	lockUuid := uuid.New()
+
+	var err error
+	for i := 0; i < retries; i++ {
+		_, err := dbi.lock(ctx, sector, read, write, lockUuid)
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Duration(waitTime) * time.Second)
```

Comment:
Ideally would be time.After in a select which would quit early when the context gets cancelled

Reply:
Sorry, I don't quite understand the above statement. Could you provide an example of what you're suggesting?

Reply:
```
select {
case <-time.After(time.Duration(waitTime) * time.Second):
case <-ctx.Done():
  return ctx.Err()
}
```

Reply:
done